scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+        return '';
+    }
+    var self = this;
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+    var type = shape.type;
+    if (type === 'structure') {
+        code += tabs(tabCount) + 'export interface ' + shapeKey + ' {\n';
+        var members = shape.members;
+        // cycle through members
+        var memberKeys = Object.keys(members);
+        memberKeys.forEach(function(memberKey) {
+            // docs
+            var member = members[memberKey];
+            if (member.documentation) {
+                code += self.generateDocString(member.documentation, tabCount + 1);
+            }
+            var required = self.checkRequired(shape.required || [], memberKey) ? '?' : '';
+            code += tabs(tabCount + 1) + memberKey + required + ': ' + member.shape + ';\n';
+        });
+        code += tabs(tabCount) + '}\n';
+    } else if (type === 'list') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = ' + shape.member.shape + '[];\n';
+    } else if (type === 'map') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = {[key: string]: ' + shape.value.shape + '};\n';
+    } else if (type === 'string' || type === 'character') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = string;\n';
+    } else if (['double', 'long', 'short', 'biginteger', 'bigdecimal', 'integer', 'float'].indexOf(type) >= 0) {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = number;\n';
+    } else if (type === 'timestamp') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Date;\n';
+    } else if (type === 'boolean') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = boolean;\n';
+    } else if (type === 'blob' || type === 'binary') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Buffer|Uint8Array|Blob|string;\n';
+    }
+    return code;
+};
+
+/**
+ * Generates a class method type for an operation.
+ */
+TSGenerator.prototype.generateTypingsFromOperations = function generateTypingsFromOperations(className, operation, tabCount) {
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+
+    var input = operation.input;
+    var output = operation.output;
+    var operationName = operation.name.charAt(0).toLowerCase() + operation.name.substring(1);
+
+    var inputShape = input ? className + '.' + input.shape : '{}';
+    var outputShape = output ? className + '.' + output.shape : '{}';
+
+    code += this.generateDocString(operation.documentation, tabCount);
+    code += tabs(tabCount) + operationName + '(params: ' + inputShape + ', callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
+    code += this.generateDocString(operation.documentation, tabCount);
+    code += tabs(tabCount) + operationName + '(callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
```

Comment:
Perhaps we could check whether the input has any required params, and if it has at least one, then don't generate this second second declaration where params is optional.